File: ios/Classes/ChannelManager.swift
import Foundation
import Flutter
@objc public class ChannelManager: NSObject {

    @objc public static let sharedInstance = ChannelManager()

    private var lock = NSLock()
    public var channel: FlutterMethodChannel?
    private var messengers = [String: BaseMessengerHandler]()

    private override init() {}

    func register(with registrar: FlutterPluginRegistrar) {
            channel = FlutterMethodChannel(name: "flutter_core_channel", binaryMessenger: registrar.messenger())
            channel?.setMethodCallHandler { [weak self] call, result in
                self?.handleFlutterMessage(call, result: result)
            }
    }

    @objc public func addAllMessengers(_ messengers: [BaseMessengerHandler]?) {
        if let messengers = messengers {
            safeExe {
                for messenger in messengers {
                    self.messengers[messenger.methodName()] = messenger
                }
            }
        }
    }

    @objc func addMessenger(_ messenger: BaseMessengerHandler?) {
        if let messenger = messenger {
            safeExe {
                self.messengers[messenger.methodName()] = messenger
            }
        }
    }

    @objc func removeMessenger(_ methodName: String?) {
        if let methodName = methodName {
            safeExe {
                self.messengers.removeValue(forKey: methodName)
            }
        }
    }

    @objc func removeAllMessengers() {
        safeExe {
            self.messengers.removeAll()
        }
    }

    public func send(_ methodName: String) {
        channel!.invokeMethod(methodName, arguments: encode(wrap(code: 0, data: nil, msg: "success")), result: nil)
    }

    @objc public func send(_ methodName: String, params: FlutterMessengerMap) {
        channel?.invokeMethod(methodName, arguments: encode(wrap(code: 0, data: params, msg: "success")), result: nil)
    }

    @objc public func send(_ methodName: String, params: FlutterMessengerMap, result: MessageResult) {
        channel!.invokeMethod(methodName, arguments: encode(wrap(code: 0, data: params, msg: "success")))
        //, result: result
    }

    private func handleFlutterMessage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let messenger = messengers[call.method], let json = call.arguments as? String else {
            result(encode(wrap(code: -999, data: nil, msg: "failed")))
            return
        }

        var decode = decode(json) ?? [:]
        let ret = messenger.didReceivedFlutterSignal(params: decode)
        result(encode(wrap(code: 0, data: ret, msg: "success")))
    }

    private func decode(_ data: Any?) -> [String: Any]? {
        guard let data = data as? String, let jsonData = data.data(using: .utf8) else {
            return nil
        }
        return try? JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any]
    }

    private func encode(_ data: [String: Any]?) -> String? {
        guard let data = data else {
            return nil
        }
        let jsonData = try? JSONSerialization.data(withJSONObject: data, options: .fragmentsAllowed)
        return String(data: jsonData ?? Data(), encoding: .utf8)
    }

    func wrap(code: Int, data: [String: Any]? = nil, msg: String = "") -> [String: Any] {
        var ret = [String: Any]()
        ret["code"] = "\(code)"
        ret["data"] = data ?? [String: Any]()
        ret["msg"] = msg
        return ret
    }

    func safeExe(block: @escaping () -> Void) {
        if Thread.isMainThread {
            block()
        } else {
            lock.lock()
            block()
            lock.unlock()
        }
    }

}

File: android/src/main/kotlin/net/allinmeta/flutter_core_channel/ChannelManager.kt
package net.allinmeta.flutter_core_channel

import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Flutter通道管理类。主要作用是
 * 1. 注册通道
 * 2. 销毁通道
 * 3. 添加消息处理器
 * 
 */

object ChannelManager: MethodChannel.MethodCallHandler {

    /// 消息通道
    private lateinit var methodChannel: MethodChannel;

    private val messengers: HashMap<String, BaseMessengerHandler> by lazy {
        val ret = HashMap<String, BaseMessengerHandler>()
        ret
    }

    /// 注册通道
    fun registerWithEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_core_channel")
        methodChannel.setMethodCallHandler(this)
    }

    /// 销毁通道
    fun destroy() {
        methodChannel.setMethodCallHandler(null)
    }

    /// 添加多个消息处理器
    @Synchronized
    fun addAllMessengers(messengers: Array<BaseMessengerHandler>) {
        for (m: BaseMessengerHandler in messengers) {
            ChannelManager.messengers.put(key = m.methodName(), value = m)
        }
    }

    /// 添加单个消息处理器
    @Synchronized
    fun addMessenger(messenger: BaseMessengerHandler) {
        messengers.put(key = messenger.methodName(), value = messenger)
    }

    /// 移除单个消息处理器
    @Synchronized
    fun removeMessenger(methodName: String) {
        messengers.remove(methodName)
    }

    /// 移除所有消息处理器
    @Synchronized
    fun removeAllMessengers() {
        messengers.clear()
    }

    /// 发送消息值Flutter
    /// route 消息路由
    /// action 目的地
    /// params 消息参数
    /// result flutter给原生的回调
    fun send(methodName: String, params: FlutterMessengerMap? = FlutterMessengerMap(), result: MessageResult? = null) {
        methodChannel.invokeMethod(
                methodName,
                encode(wrap(msg = "success", data = params)),
                FlutterSendMessageCallBack(result)
        )
    }

    /// 接收Flutter传递的消息
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        var messenger: BaseMessengerHandler? = messengers[call.method]
        if (messenger == null) {
            Log.e("KooChannel","在messengers<${messengers.keys}>中未找到${call.method}对应的messenger，确保调用前已经注册")
            result.success(encode(wrap(msg = "failed", code = -999)))
        }else{
            val params: HashMap<String, Any> = decode((call.arguments as String))
            val action: String = params.remove("action") as String
            var ret: FlutterMessengerMap? = messenger.didReceivedFlutterSignal(params, action)
            result.success(encode(wrap(data = ret, msg = "success")))
        }

    }


}